### ISSUE TITLE
Nested generic extensions require `where` clauses instead of <…>

### ISSUE THREAD

**Issue by JessyCatterwaul** [apple/swift#63866]
Given 

```swift
enum Outer<O> {
  struct Inner<I> { }
}
```

This form compiles:

```swift
extension Outer.Inner where I == Outer { }
```

But this produces the error, "Reference to generic type 'Outer' requires arguments in <...>".
```swift
extension Outer.Inner<Outer> { }
```

**Comment by AnthonyLatsis** [comment]
I am almost certain this is intentional as per SE-0361. The extended type itself is not and was never resolved in the context of the extended type declaration, unlike the `where` clause:

```swift
struct S<T> {
  typealias A = Int
}
extension S where T == A {}
extension S<A> {} // error: cannot find type 'A' in scope
```

Doing so is a source-breaking and, in my opinion, arguable change that would need to go through the Swift evolution process. For example, would this mean the generic parameters could be referenced in the extended type as well?

```swift
extension Array<Element> {} // Okay...
extension Array<Element?> {} // What?
``` 

cc @hborla @slavapestov for more viewpoints.

**Comment by slavapestov** [comment]
You either have to specify all the generic arguments, or none of them. So this works:

```swift
extension Outer<Int>.Inner<String>
```

I'm not sure if we consider this a bug or working as intended though.

**Comment by JessyCatterwaul** [comment]
> You either have to specify all the generic arguments, or none of them.

This is not true. Testing to make sure, I realized that `extension Outer.Inner<Outer>` not compiling is a regression. It works  with Swift 5.7.

_I'm working on creating another example which more closely mimics my use case; will update here._

**Comment by JessyCatterwaul** [comment]
Here's the simplest I've been able to whittle it down to. The last line generates 

- Swift 5.8: `Reference to generic type 'Outer' requires arguments in <...>`
- Swift 5.7: `nonzero exit code`

```swift
import SwiftUI

class Outer<O>: ObservableObject {
  enum Inner<I: wrappedValue<Outer>> { }
}

protocol wrappedValue<WrappedValue> {
  associatedtype WrappedValue
  var wrappedValue: WrappedValue { get }
}

extension ObservedObject: wrappedValue { }

extension Outer.Inner where I == ObservedObject<Outer> { }
extension Outer.Inner<ObservedObject<Outer>> { }
```
